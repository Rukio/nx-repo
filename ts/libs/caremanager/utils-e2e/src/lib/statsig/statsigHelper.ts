import { Statsig } from '@*company-data-covered*/cypress-shared';
import { StatsigEnvironment, StatsigOptions } from 'statsig-js';

type StatsigTier = StatsigEnvironment['tier'];

const statsigClientKey = Cypress.env('STATSIG_CLIENT_KEY');
const statsigUser = null;
const statsigTier: StatsigTier =
  (Cypress.env('STATSIG_ENV') as StatsigTier) ?? 'development';
const statsigOptions: StatsigOptions = {
  localMode: true,
  environment: {
    tier: statsigTier,
  },
};

const experimentsList = {};
const featureGatesList = {
  maintenanceMode: 'caremanager_maintenance_mode',
  visitsV1: 'caremanager_visits_v1',
};

export const statsigHelper = new Statsig({
  options: {
    user: statsigUser,
    clientKey: statsigClientKey,
    ...statsigOptions,
  },
  featureGatesList,
  experimentsList,
});

statsigHelper.initialize();
