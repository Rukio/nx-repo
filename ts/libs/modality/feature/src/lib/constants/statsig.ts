export const INSURANCE_RESTRUCTURE_FEATURE_GATE_NAME = 'insurance_restructure';
