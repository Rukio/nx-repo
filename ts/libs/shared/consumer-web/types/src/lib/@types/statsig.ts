export enum StatsigEvents {
  EVENT = 'generic_event',
  REQUEST_START_EVENT = 'web_request_start',
  REQUEST_CONTACT_INFO_EVENT = 'web_request_contact_info',
  REQUEST_PREFERRED_TIME_WINDOW = 'web_preferred_time_window',
  REQUEST_ADDRESS_EVENT = 'web_request_address',
  REQUEST_REQUESTOR_TYPE_EVENT = 'web_request_requestor_type',
  REQUEST_PERSONAL_INFO_EVENT = 'web_request_personal_info',
  REQUEST_SYMPTOMS_EVENT = 'web_request_symptoms',
  REQUEST_SUBMITTED_EVENT = 'web_request_submitted',
  REQUEST_HOW_IT_WORKS_EVENT = 'web_request_how_it_works',
  REQUEST_SYMPTOMS_SELECTION_EVENT = 'web_request_symptom_selection',
  ADD_REQUIRED_INSURANCE_PAGE_EVENT = 'add_insurance_pageview',
  REQUEST_COMPLETE_CONFIRMATION_OPTIONAL_INSURANCE_EVENT = 'request_complete_confirmation_insurance_optional',
  ADD_INSURANCE_PHOTO_EVENT = 'add_insurance_photo_attempt',
  ADD_INSURANCE_MANUALLY_EVENT = 'add_insurance_manual_attempt',
  NO_INSURANCE_EVENT = 'add_insurance_no_insurance',
  ADD_INSURANCE_PHOTO_SUCCESS_EVENT = 'add_insurance_photo_success',
  ADD_INSURANCE_MANUALLY_SUCCESS_EVENT = 'add_insurance_manual_success',
  ADD_INSURANCE_MANUALLY_SAVE_EVENT = 'add_insurance_manual_save',
  ADD_INSURANCE_PHOTO_SUBMIT_EVENT = 'add_insurance_photo_submit_images',
  ADD_INSURANCE_SUBMIT_EVENT = 'add_insurance_submit_visit_request',
  ADD_INSURANCE_SUBMIT_SUCCESS_EVENT = 'request_complete_confirmation_insurance_added',
  NO_INSURANCE_SUBMIT_SUCCESS_EVENT = 'request_complete_confirmation_no_insurance',
}

export enum StatsigCategories {
  FORM_ERRORS = 'form errors',
  ERRORS = 'errors',
  ERROR_404 = '404',
  ERROR_500 = '500',
  BROKEN_IMAGE = 'broken image',
}
export enum StatsigActions {
  ADDED_CONTACT_INFO = 'added contact info',
  ADDED_PREFERRED_TIME_WINDOW = 'added preferred time window',
  ADDED_ADDRESS = 'added address',
  ADDED_REQUESTOR_TYPE = 'added requestor type',
  ADDED_PERSONAL_INFO = 'added personal info',
  ADDED_SYMPTOMS = 'added symptoms',
  SUBMIT_REQUEST = 'submitted care request',
  COMPLETED_CARE_REQUEST = 'completed care request',
  WEB_REQUEST_TYPE = 'web',
  MOBILE_WEB_REQUEST_TYPE = 'mobile_web',
}
export enum StatsigLabels {
  PAGE_URL = 'Page URL',
  DEFAULT_LABEL = '-',
  SHARE = 'share',
  IMAGE = 'image',
  CLOSE = 'close',
  PREVIOUS = 'previous',
  NEXT = 'next',
}
export enum StatsigPageCategory {
  REQUEST_CONTACT_INFO = 'contact_info',
  REQUEST_PREFERRED_TIME_WINDOW = 'request_preferred_time_window',
  REQUEST_ADDRESS = 'request_address',
  REQUEST_CARE_FOR = 'request_care_for',
  REQUEST_HELP = 'request_help',
  REQUEST_DETAILS = 'request_personal_info',
  REQUEST_INSURNACE = 'request_insurance',
  REQUEST_HOW_IT_WORKS = 'request_how_it_works',
  ERROR_404 = '404',
  ERROR_500 = '500',
  ERROR = 'error',
}
