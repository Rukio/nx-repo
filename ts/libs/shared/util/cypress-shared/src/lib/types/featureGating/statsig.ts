import { StatsigOptions, StatsigUser } from 'statsig-js';

export declare namespace StatsigOverrides {
  type ExperimentKey = string;
  type ExperimentOverrideParamMap = Record<
    ExperimentKey,
    Record<string, unknown>
  >;

  type FeatureGateKey = string;
  type FeatureGateOverrideParamMap = Record<FeatureGateKey, boolean>;
}

export declare namespace StatsigHelper {
  type ExperimentKey = StatsigOverrides.ExperimentKey;
  type FeatureGateKey = StatsigOverrides.FeatureGateKey;

  type ExperimentList = Record<ExperimentKey, string>;
  type FeatureGateList = Record<FeatureGateKey, string>;

  type Options = {
    clientKey: string;
    user: StatsigUser | null;
  } & StatsigOptions;

  type Initialize<E extends ExperimentList, F extends FeatureGateList> = {
    options: Options;
    experimentsList: E;
    featureGatesList: F;
  };

  type FeatureGateOverrideParamMap =
    StatsigOverrides.FeatureGateOverrideParamMap;
  type ExperimentOverrideParamMap = StatsigOverrides.ExperimentOverrideParamMap;
}
