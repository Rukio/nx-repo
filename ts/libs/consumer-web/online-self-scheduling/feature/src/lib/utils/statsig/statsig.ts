import statsig from 'statsig-js';

export const RISK_STRAT_MAPPING_FOR_OSS_CONFIG_NAME =
  'risk_strat_mapping_for_oss';

export const ONBOARDING_ACUITY_SEGMENTATION_MARKETS_CONFIG_NAME =
  'onboarding_acuity_segmentation_markets';

export const CHANNEL_ITEM_TO_MEDIUM_MAPPING_CONFIG_NAME =
  'channel_item_to_medium_mapping';

export enum StructuredSymptomCallTo {
  Dispatcher = 'Dispatcher',
  Screener = 'Screener',
  NoCall = 'No Call',
}

export type StructuredSymptom = {
  friendly_name: string;
  is_oss_eligible: boolean;
  legacy_rs_protocol: string;
  route_call_to: StructuredSymptomCallTo;
  legacy_rs_protocol_id: number;
};

export type AcuitySegmentationMarketClassifications = {
  markets: string[];
  classifications: number[];
};

export interface ChannelItem {
  medium?: string;
  channel_item_name: string;
  channel_item_id: number;
}

export const getStructuredSymptoms = () => {
  return statsig
    .getConfig(RISK_STRAT_MAPPING_FOR_OSS_CONFIG_NAME)
    .get<StructuredSymptom[]>('structured_symptoms', []);
};

export const getStructuredSymptomBySelectedSymptoms = (
  selectedSymptoms?: string
): StructuredSymptom | null => {
  if (!selectedSymptoms) {
    return null;
  }
  const structuredSymptoms = getStructuredSymptoms();

  return (
    structuredSymptoms.find(
      (structuredSymptom) =>
        structuredSymptom.friendly_name === selectedSymptoms
    ) ?? null
  );
};

export const getAcuitySegmentationMarketClassifications = () => {
  return statsig
    .getConfig(ONBOARDING_ACUITY_SEGMENTATION_MARKETS_CONFIG_NAME)
    .get<AcuitySegmentationMarketClassifications>(
      'acuity_segmentation_launched_markets',
      {
        markets: [],
        classifications: [],
      }
    );
};

export const getChannelItems = () => {
  return statsig
    .getConfig(CHANNEL_ITEM_TO_MEDIUM_MAPPING_CONFIG_NAME)
    .get<ChannelItem[]>('channel_items', []);
};

export const getDefaultChannelItem = () => {
  return statsig
    .getConfig(CHANNEL_ITEM_TO_MEDIUM_MAPPING_CONFIG_NAME)
    .get<ChannelItem | null>('default_channel_item', null);
};
