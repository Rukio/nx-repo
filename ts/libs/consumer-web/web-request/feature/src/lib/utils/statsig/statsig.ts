import { logError } from '@*company-data-covered*/shared/datadog/util';
import statsig from 'statsig-js';

export interface StatsigLogEvent {
  eventName: string;
  value?: string | number | null;
  metadata?: Record<string, string> | null;
}

export const getStatsigStableId = () =>
  localStorage.STATSIG_LOCAL_STORAGE_STABLE_ID;

export const StatsigExperiments = {
  web_request_schedule_time_window_page: {
    name: 'web_request_schedule_time_window_page',
    params: {
      render_time_window_page: 'render_time_window_page',
    },
  },
};

export const getMarketNameFromHowItWorksConfig = (
  marketAbbreviation: string
) => {
  try {
    const howItWorksMarketsConfig = statsig.getConfig(
      'web_request_how_it_works_markets'
    );
    const howItWorksMarkets = howItWorksMarketsConfig.get<
      {
        abbreviation: string;
        name: string;
      }[]
    >('how_it_works_markets', []);

    const marketMatch = howItWorksMarkets.find(
      (market) =>
        market.abbreviation.toLowerCase() === marketAbbreviation.toLowerCase()
    );
    if (!marketMatch) {
      return null;
    }

    return marketMatch.name;
  } catch (e: unknown) {
    logError(
      `[Statsig] error occured while getting web_request_how_it_works_markets config: ${e}`
    );

    return null;
  }
};

export const getSymptomsFromStructuredSymptomsConfig = () => {
  try {
    const structuredSymptomsConfig = statsig.getConfig(
      'web_request_structured_symptoms'
    );
    if (!structuredSymptomsConfig) {
      return [];
    }

    return structuredSymptomsConfig.get<string[]>('structured_symptoms', []);
  } catch (e: unknown) {
    logError(
      `[Statsig] error occured while getting web_request_structured_symptoms config: ${e}`
    );

    return [];
  }
};
