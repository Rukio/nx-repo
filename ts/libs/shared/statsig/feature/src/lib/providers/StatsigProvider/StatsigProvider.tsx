import { ReactNode, useEffect, useState } from 'react';
import statsig, { StatsigUser, StatsigOptions } from 'statsig-js';

/**
 * Properties required to use the StatsigProvider
 */
export type StatsigProviderProps = {
  children: ReactNode;

  /** A client Statsig key from the Statsig Console */
  clientKey: string;

  /** A Statsig User object */
  user?: StatsigUser;

  /** A Statsig Options object */
  options?: StatsigOptions;

  /** A loading component to render if waitForInitialization is set to true */
  loadingComponent?: ReactNode;

  /**
   * Waits for the Statsig SDK to initialize before rendering children. It is recommended to set this to true so that Statsig will be available for rendered child components that might rely on it.
   *
   * @default true
   */
  waitForInitialization?: boolean;

  /** Should shutdown Statsig SDK on component unmount */
  shutdownOnUnmount?: boolean;
  onInitSuccess?: () => void;
  onInitFailure?: (error: unknown) => void;
};

/**
 * The StatsigProvider is the wrapper for the entry point of the app.
 * It initializes the Statsig SDK with the option of waiting for initialization
 *
 * @example
 * ```typescript
 *
 * const ExampleComponent = (props: ComponentProps) => {
 *   return (
 *    <StatsigProvider
 *      clientKey="client-XXX"
 *      loadingComponent={<Box>Loading...</Box>}
 *    >
 *      <Box>Example</Box>
 *    </StatsigProvider>
 *   );
 * };
 * ```
 */
export const StatsigProvider: React.FC<StatsigProviderProps> = ({
  children,
  clientKey,
  user,
  options,
  loadingComponent,
  waitForInitialization = true,
  shutdownOnUnmount,
  onInitSuccess,
  onInitFailure,
}) => {
  const [initiated, setInitiated] = useState(false);

  useEffect(() => {
    if (!initiated) {
      statsig
        .initialize(clientKey, user, options)
        .then(() => {
          onInitSuccess?.();
        })
        .catch((error) => {
          onInitFailure?.(error);
        })
        .finally(() => {
          setInitiated(true);
        });
    }
  }, [initiated, clientKey, options, user, onInitSuccess, onInitFailure]);

  useEffect(() => {
    return () => {
      if (shutdownOnUnmount) {
        statsig.shutdown();
      }
    };
  }, [shutdownOnUnmount]);

  if (!initiated && waitForInitialization) {
    return <>{loadingComponent}</>;
  }

  return <>{children}</>;
};
