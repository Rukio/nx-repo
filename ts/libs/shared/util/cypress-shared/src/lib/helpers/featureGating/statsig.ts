import statsig, { DynamicConfig } from 'statsig-js';
import { StatsigHelper } from '../../types/featureGating/statsig';

export class Statsig<
  E extends StatsigHelper.ExperimentList,
  F extends StatsigHelper.FeatureGateList
> {
  /** Object of Experiments returned from Statsig, with each value a method that returns the current experiment setting */
  public experiments: Record<StatsigHelper.ExperimentKey, () => DynamicConfig>;

  /** Object of FeatureGates returned from Statsig, with each value a method that returns the current feature gate setting */
  public featureGates: Record<StatsigHelper.FeatureGateKey, () => boolean>;

  private options: StatsigHelper.Options;
  private experimentsList: Record<StatsigHelper.ExperimentKey, string>;
  private featureGatesList: Record<StatsigHelper.FeatureGateKey, string>;

  /**
   * Create a Statsig object.
   * @param {StatsigHelper.Initialize<E, F>} params - Constructor params
   * @param {Options} params.options - Initialization option params
   * @param {string} params.options.clientKey - Valid 'Client API Key' in Statsig for the *company-data-covered* organization
   * @param {StatsigUser | null} params.options.user - Specific user to initialize the client
   * @param {StatsigOptions} params.options.options - Additional param to change the default behavior of statsig.initialize.
   * @param {Record<ExperimentKey, string>} params.experimentsList - Object of experiments to return from Statsig, with each value a real experiment ID
   * @param {Record<FeatureGateKey, string>} params.featureGatesList - Object of feature gates to return from Statsig, with each value a real feature gate ID
   */
  constructor({
    options,
    experimentsList,
    featureGatesList,
  }: StatsigHelper.Initialize<E, F>) {
    this.options = options;

    this.experimentsList = experimentsList;
    this.experiments = this.getExperiments();

    this.featureGatesList = featureGatesList;
    this.featureGates = this.getFeatureGates();
  }

  public async initialize() {
    try {
      const { clientKey, user, environment, ...rest } = this.options;

      Cypress.log({
        displayName: 'Starting Statsig Initialization',
        message: `Tier: ${environment?.tier}`,
      });

      await statsig.initialize(clientKey, user, rest);

      Cypress.log({
        displayName: 'Finished Statsig Initialization',
      });
    } catch (error) {
      let message = 'Unknown Error';
      if (error instanceof Error) {
        message = error.message;
      }

      Cypress.log({
        displayName: 'Starting Statsig Initialization Failed',
        message: `Error: ${message}`,
      });
    }
  }

  /* Experiments */
  private getExperiment(name: string): DynamicConfig {
    return statsig.getExperiment(name);
  }

  private getExperiments() {
    const result = {} as typeof this.experiments;
    if (this.experimentsList) {
      const experimentKeys = Object.keys(this.experimentsList);

      for (const key of experimentKeys) {
        result[key] = () => this.getExperiment(this.experimentsList[key]);
      }
    }

    return result;
  }

  /**
   * Override an experiment
   * @param {StatsigHelper.ExperimentKey} key - statsig experiment to override
   * @param {Partial<StatsigHelper.ExperimentOverrideParamMap[StatsigHelper.ExperimentKey]} params - map of param values to use for the override
   * @param {boolean} reload - whether or not to reload the page after overriding
   */
  public overrideExperiment<K extends StatsigHelper.ExperimentKey>(
    key: K,
    params: Partial<StatsigHelper.ExperimentOverrideParamMap[K]>,
    reload = false
  ) {
    const experimentName = this.experimentsList[key];

    Cypress.log({
      displayName: 'Overriding Statsig Config',
      message: `Name: ${experimentName}, Values: ${JSON.stringify(params)}`,
    });

    statsig.overrideConfig(experimentName, params);

    if (reload) {
      cy.reload();
    }
  }

  /**
   * Override multiple experiments
   * @param {} overrides - override params
   * @param {StatsigHelper.ExperimentKey} overrides.key - statsig experiment to override
   * @param {Partial<StatsigHelper.ExperimentOverrideParamMap[StatsigHelper.ExperimentKey]} overrides.params - map of param values to use for the override
   * @param {boolean} reload - whether or not to reload the page after overriding
   */
  public overrideExperiments<K extends StatsigHelper.ExperimentKey>(
    overrides: {
      key: K;
      params: Partial<StatsigHelper.ExperimentOverrideParamMap[K]>;
    }[],
    reload = true
  ) {
    for (const { key, params } of overrides) {
      this.overrideExperiment(key, params, false);
    }

    if (reload) {
      cy.reload();
    }
  }

  /* Feature Gates */
  private getFeatureGate(name: string): boolean {
    return statsig.checkGate(name);
  }

  private getFeatureGates() {
    const result = {} as typeof this.featureGates;
    if (this.featureGatesList) {
      const featureGateKeys = Object.keys(this.featureGatesList);

      for (const key of featureGateKeys) {
        result[key] = () => this.getFeatureGate(this.featureGatesList[key]);
      }
    }

    return result;
  }

  /**
   * Override a feature gate
   * @param {StatsigHelper.FeatureGateKey} key - statsig feature gate to override
   * @param {boolean} value - value to use for the override
   * @param {boolean} reload - whether or not to reload the page after overriding
   */
  public overrideFeatureGate<K extends StatsigHelper.FeatureGateKey>(
    key: K,
    value: boolean,
    reload = false
  ) {
    const gateName = this.featureGatesList[key];

    Cypress.log({
      displayName: 'Overriding Statsig Feature Gate',
      message: `Name: ${gateName}, Values: ${value}`,
    });

    statsig.overrideGate(gateName, value);

    if (reload) {
      cy.reload();
    }
  }

  /**
   * Override multiple feature gates
   * @param {} overrides - override params
   * @param {StatsigHelper.FeatureGateKey} overrides.key - statsig feature gate to override
   * @param {Partial<StatsigHelper.FeatureGateOverrideParamMap[StatsigHelper.FeatureGateKey]} overrides.params - map of param values to use for the override
   * @param {boolean} reload - whether or not to reload the page after overriding
   */
  public overrideFeatureGates<K extends StatsigHelper.FeatureGateKey>(
    overrides: {
      key: K;
      params: Partial<StatsigHelper.FeatureGateOverrideParamMap[K]>;
    }[],
    reload = true
  ) {
    for (const { key, params } of overrides) {
      this.overrideFeatureGate(key, params, false);
    }

    if (reload) {
      cy.reload();
    }
  }
}
