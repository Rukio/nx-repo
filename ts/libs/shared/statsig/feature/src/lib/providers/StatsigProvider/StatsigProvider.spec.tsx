import { render, screen, waitFor } from '@testing-library/react';
import { StatsigProvider } from './StatsigProvider';
import statsig from 'statsig-js';

vi.mock('statsig-js', async () => {
  const actual = await vi.importActual<typeof import('statsig-js')>(
    'statsig-js'
  );

  return {
    ...actual,
    default: {
      ...actual.default,
      initialize: vi.fn().mockResolvedValue({}),
      shutdown: vi.fn(),
    },
  };
});

const mockStatsigInitialize = vi.mocked(statsig.initialize);
const mockStatsigShutdown = vi.mocked(statsig.shutdown);

export const mockContentProps = {
  dataTestId: 'mock-content-component',
  textContent: 'Content',
};

export const mockLoadingComponentProps = {
  dataTestId: 'mock-loading-component',
  textContent: '...Loading',
};

export const MockLoadingComponent = () => (
  <div data-testid={mockLoadingComponentProps.dataTestId}>
    {mockLoadingComponentProps.textContent}
  </div>
);

export const MockContentComponent = () => (
  <div data-testid={mockContentProps.dataTestId}>
    {mockContentProps.textContent}
  </div>
);

describe('StatsigProvider', () => {
  beforeEach(() => {
    vi.resetModules();
    vi.clearAllMocks();
  });

  it('should wait for init and render loading component correctly and render component after init', async () => {
    const { asFragment } = render(
      <StatsigProvider
        clientKey="client-XXX"
        loadingComponent={<MockLoadingComponent />}
      >
        <MockContentComponent />
      </StatsigProvider>
    );

    expect(asFragment()).toMatchSnapshot();

    const loadingComponent = screen.getByTestId(
      mockLoadingComponentProps.dataTestId
    );
    expect(loadingComponent.textContent).toContain(
      mockLoadingComponentProps.textContent
    );

    const contentComponent = await screen.findByTestId(
      mockContentProps.dataTestId
    );

    expect(contentComponent.textContent).toContain(
      mockContentProps.textContent
    );
  });

  it('should render content after statsig init failure', async () => {
    mockStatsigInitialize.mockRejectedValue(new Error());
    render(
      <StatsigProvider
        clientKey="client-XXX"
        loadingComponent={<MockLoadingComponent />}
      >
        <MockContentComponent />
      </StatsigProvider>
    );

    const loadingComponent = screen.getByTestId(
      mockLoadingComponentProps.dataTestId
    );
    expect(loadingComponent.textContent).toContain(
      mockLoadingComponentProps.textContent
    );

    const contentComponent = await screen.findByTestId(
      mockContentProps.dataTestId
    );
    expect(contentComponent.textContent).toContain('Content');
  });

  it('should render content after statsig init success', async () => {
    mockStatsigInitialize.mockResolvedValue();
    render(
      <StatsigProvider
        clientKey="client-XXX"
        loadingComponent={<MockLoadingComponent />}
      >
        <MockContentComponent />
      </StatsigProvider>
    );

    const loadingComponent = screen.getByTestId(
      mockLoadingComponentProps.dataTestId
    );
    expect(loadingComponent.textContent).toContain(
      mockLoadingComponentProps.textContent
    );

    const contentComponent = await screen.findByTestId(
      mockContentProps.dataTestId
    );
    expect(contentComponent.textContent).toContain('Content');
  });

  it('should render content immediately when waitForInitialization is falsy', () => {
    render(
      <StatsigProvider
        clientKey="client-XXX"
        loadingComponent={<MockLoadingComponent />}
        waitForInitialization={false}
      >
        <MockContentComponent />
      </StatsigProvider>
    );

    const loadingComponent = screen.queryByTestId(
      mockLoadingComponentProps.dataTestId
    );
    expect(loadingComponent).toBeNull();

    const contentComponent = screen.getByTestId(mockContentProps.dataTestId);
    expect(contentComponent.textContent).toContain('Content');
  });

  it('should call shutdown method on umount if shutdownOnUnmount prop is truthy', async () => {
    mockStatsigInitialize.mockResolvedValue();
    const { unmount } = render(
      <StatsigProvider
        clientKey="client-XXX"
        loadingComponent={<MockLoadingComponent />}
        shutdownOnUnmount
      >
        <MockContentComponent />
      </StatsigProvider>
    );

    const loadingComponent = screen.getByTestId(
      mockLoadingComponentProps.dataTestId
    );
    expect(loadingComponent.textContent).toContain(
      mockLoadingComponentProps.textContent
    );

    const contentComponent = await screen.findByTestId(
      mockContentProps.dataTestId
    );
    expect(contentComponent.textContent).toContain('Content');

    unmount();

    await waitFor(() => {
      expect(mockStatsigShutdown).toBeCalled();
    });
  });

  it('should not call shutdown method on umount if shutdownOnUnmount prop is falsy', async () => {
    mockStatsigInitialize.mockResolvedValue();
    const { unmount } = render(
      <StatsigProvider
        clientKey="client-XXX"
        loadingComponent={<MockLoadingComponent />}
        shutdownOnUnmount={false}
      >
        <MockContentComponent />
      </StatsigProvider>
    );

    const loadingComponent = screen.getByTestId(
      mockLoadingComponentProps.dataTestId
    );
    expect(loadingComponent.textContent).toContain(
      mockLoadingComponentProps.textContent
    );

    const contentComponent = await screen.findByTestId(
      mockContentProps.dataTestId
    );
    expect(contentComponent.textContent).toContain('Content');

    unmount();

    await waitFor(() => {
      expect(mockStatsigShutdown).not.toBeCalled();
    });
  });
});
