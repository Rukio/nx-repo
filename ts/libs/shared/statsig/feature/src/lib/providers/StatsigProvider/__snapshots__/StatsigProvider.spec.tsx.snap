// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StatsigProvider > should wait for init and render loading component correctly and render component after init 1`] = `
<DocumentFragment>
  <div
    data-testid="mock-loading-component"
  >
    ...Loading
  </div>
</DocumentFragment>
`;
